app/modules/ws2812_effects.c

```diff
@@ -0,0 +1,1074 @@
+
+
+
+typedef struct {
+  ws2812_buffer *buffer;
+  uint32_t mode_delay;
+  uint32_t counter_mode_call;
+  uint32_t counter_mode_step;
+  uint8_t mode_color_index;
+  uint8_t speed;
+  uint8_t brightness;
+  os_timer_t os_t;
+  uint8_t running;
+  uint8_t effect_type;
+  uint8_t color[4];
+  int effect_int_param1;
+} ws2812_effects;
+
+
+
+enum ws2812_effects_type {
+  WS2812_EFFECT_STATIC,
+  WS2812_EFFECT_BLINK,
+  WS2812_EFFECT_GRADIENT,
+  WS2812_EFFECT_GRADIENT_RGB,
+  WS2812_EFFECT_RANDOM_COLOR,
+  WS2812_EFFECT_RAINBOW,
+  WS2812_EFFECT_RAINBOW_CYCLE,
+  WS2812_EFFECT_FLICKER,
+  WS2812_EFFECT_FIRE_FLICKER,
+  WS2812_EFFECT_FIRE_FLICKER_SOFT,
+  WS2812_EFFECT_FIRE_FLICKER_INTENSE,
+  WS2812_EFFECT_HALLOWEEN,
+  WS2812_EFFECT_CIRCUS_COMBUSTUS,
+  WS2812_EFFECT_LARSON_SCANNER,
+  WS2812_EFFECT_CYCLE,
+  WS2812_EFFECT_COLOR_WIPE,
+  WS2812_EFFECT_RANDOM_DOT
+};
+
+
+static ws2812_effects *state;
+
+
+//-----------------
+// UTILITY METHODS
+//-----------------
+
+
+static int ws2812_write(ws2812_buffer* buffer) {
+  size_t length1, length2;
+  const char *buffer1, *buffer2;
+
+  buffer1 = 0;
+  length1 = 0;
+
+  buffer1 = buffer->values;
+  length1 = buffer->colorsPerLed*buffer->size;
+
+  buffer2 = 0;
+  length2 = 0;
+
+  // Send the buffers
+  ws2812_write_data(buffer1, length1, buffer2, length2);
+
+  return 0;
+}
+
+
+static int ws2812_set_pixel(int pixel, uint32_t color) {
+  ws2812_buffer * buffer = state->buffer;
+  uint8_t g = ((color & 0x00FF0000) >> 16);
+  uint8_t r = ((color & 0x0000FF00) >> 8);
+  uint8_t b = (color & 0x000000FF);
+  uint8_t w = buffer->colorsPerLed  == 4 ? ((color & 0xFF000000) >> 24) : 0;
+
+  int offset = pixel * buffer->colorsPerLed;
+  buffer->values[offset] = g;
+  buffer->values[offset+1] = r;
+  buffer->values[offset+2] = b;
+  if (buffer->colorsPerLed == 4) {
+    buffer->values[offset+3] = w;
+  }
+
+  return 0;
+}
+
+
+/*
+* Returns a new, random color wheel index with a minimum distance of 42 from pos.
+*/
+static uint8_t get_random_wheel_index(uint8_t pos)
+{
+  uint8_t r = 0;
+  uint8_t x = 0;
+  uint8_t y = 0;
+  uint8_t d = 0;
+
+  while(d < 42) {
+    r = rand() % 360;
+    x = abs(pos - r);
+    y = 360 - x;
+    d = min(x, y);
+  }
+
+  return r;
+}
+
+//-----------------
+// EFFECTS LIBRARY
+//-----------------
+
+/**
+* initialized ws2812_effects with the buffer to use
+*/
+static int ws2812_effects_init(lua_State *L) {
+  ws2812_buffer * buffer = (ws2812_buffer*)luaL_checkudata(L, 1, "ws2812.buffer");
+  luaL_argcheck(L, buffer != NULL, 1, "no valid buffer provided");
+  // get rid of old state
+  if (state != NULL) {
+    os_free((void *) state);
+  }
+  // Allocate memory and set all to zero
+  size_t size = sizeof(ws2812_effects) + buffer->colorsPerLed*sizeof(uint8_t);
+  state = (ws2812_effects *) os_zalloc(size);
+  // initialize
+  state->speed = SPEED_DEFAULT;
+  state->mode_delay = DELAY_DEFAULT;
+  state->brightness = BRIGHTNESS_DEFAULT;
+  state->buffer = buffer;
```

Comment:
You'll need to [`luaL_ref()`](http://pgl.yoyo.org/luai/i/luaL_ref) the buffer object to prevent it from being garbage collected after this function exists.